library/std/src/net/ip.rs

```diff
@@ -1135,9 +1136,10 @@ impl Ipv6Addr {
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc00a, 0x2ff).is_unspecified(), false);
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0).is_unspecified(), true);
     /// ```
+    #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     #[stable(since = "1.7.0", feature = "ip_17")]
-    pub fn is_unspecified(&self) -> bool {
-        self.segments() == [0, 0, 0, 0, 0, 0, 0, 0]
+    pub const fn is_unspecified(&self) -> bool {
+        matches!(self.segments(), [0, 0, 0, 0, 0, 0, 0, 0])
```

Comment:
As a general rule, `matches!` won't optimize as well as comparison to an array I don't think the performance of `is_unspecified` and `is_loopback` is terribly important for real code, so this is fine for now. It would be good to look into constifying slice equality, making `memcmp` callable in a const context, or generating better code for array patterns in `match` statements.

Reply:
Yeah, which is a shame. I was looking into this for the `Ipv4Addr` methods as well, most of which can be written very elegantly using `matches!`. If performance is a concern, the same implementation could be used as in [`Ipv4Addr::is_global`](https://github.com/rust-lang/rust/blob/58d5ce4929ff4aee3d20973080e0391eb775f3d6/library/std/src/net/ip.rs#L548-L550):
```rust
u128::from_be_bytes(self.octets()) == 0
```

Reply:
Yeah, let's do it that way. Instead of comparing with an integer literal, however, you should convert the `LOOPBACK` and `UNSPECIFIED` associated constants to a `u128` and compare with them.